src/unix/signal.c

```diff
@@ -1,24 +1,765 @@
+        return SIGKILL;
+    if (masked & mask_from_sig(SIGSTOP))
+        return SIGSTOP;
+    return lsb(masked) + 1;
+}
+
+static inline u64 normalize_mask(u64 mask)
+{
+    return mask & ~(mask_from_sig(SIGKILL) | mask_from_sig(SIGSTOP));
+}
+
+static inline u64 sigstate_get_mask(sigstate ss)
+{
+    return normalize_mask(ss->mask);
+}
+
+static inline void sigstate_set_mask(sigstate ss, u64 mask)
+{
+    ss->mask = normalize_mask(mask);
+}
+
+static inline void sigstate_block(sigstate ss, u64 mask)
+{
+    ss->mask |= normalize_mask(mask);
+}
+
+static inline void sigstate_unblock(sigstate ss, u64 mask)
+{
+    ss->mask &= ~mask;
+}
+
+static inline u64 sigstate_get_pending(sigstate ss)
+{
+    return ss->pending;
+}
+
+static inline u64 sigstate_get_ignored(sigstate ss)
+{
+    return ss->ignored;
+}
+
+static inline u64 sigstate_set_ignored(sigstate ss, u64 mask)
+{
+    return ss->ignored = normalize_mask(mask);
+}
+
+static inline u64 sigstate_get_pending_masked(sigstate ss, u64 aux_mask)
+{
+    return ss->pending & ~(sigstate_get_mask(ss) | sigstate_get_ignored(ss) | aux_mask);
+}
+
+static inline void sigstate_set_pending(sigstate ss, int sig)
+{
+    ss->pending |= mask_from_sig(sig);
+}
+
+static inline boolean sigstate_is_pending(sigstate ss, int sig)
+{
+    return (ss->pending & mask_from_sig(sig)) != 0;
+}
+
+static inline list sigstate_get_sighead(sigstate ss, int signum)
+{
+    assert(signum > 0 && signum <= 64);
+    return &ss->heads[signum - 1];
+}
+
+static inline void sigstate_restore(sigstate ss)
+{
+    ss->mask = ss->saved;
+    ss->saved = 0;
+}
+
+static queued_signal sigstate_dequeue_signal(sigstate ss, u64 aux_mask)
+{
+    sig_debug("sigstate %p\n", ss);
+    u64 masked = sigstate_get_pending_masked(ss, aux_mask);
+    if (masked == 0)
+        return INVALID_ADDRESS;
+
+    int signum = select_signal_from_masked(masked);
+    if (!signum)
+        return INVALID_ADDRESS;
+    sigaction sa = get_sigaction(signum);
+    u64 sigword = mask_from_sig(signum);
+    if (ss->saved == 0)      /* rt_sigsuspend may provide one */
+        ss->saved = ss->mask;
+    ss->mask |= sigword | sa->sa_mask.sig[0];
+    sig_debug("sig %d, now sigsaved 0x%lx, sigmask 0x%lx\n",
+              signum, ss->saved, ss->mask);
+
+    /* dequeue siginfo */
+    list head = sigstate_get_sighead(ss, signum);
+    list l = list_get_next(head);
+    assert(l);
+    queued_signal qs = struct_from_list(l, queued_signal, l);
+    list_delete(l);
+    if (list_empty(head))
+        ss->pending &= ~sigword;
+    return qs;
+}
+
+void init_sigstate(sigstate ss)
+{
+    ss->pending = 0;
+    ss->mask = 0;
+    ss->saved = 0;
+    ss->ignored = mask_from_sig(SIGCHLD) | mask_from_sig(SIGURG) | mask_from_sig(SIGWINCH);
+
+    for(int i = 0; i < NSIG; i += 4) {
+        list_init(&ss->heads[i]);
+        list_init(&ss->heads[i + 1]);
+        list_init(&ss->heads[i + 2]);
+        list_init(&ss->heads[i + 3]);
+    }
```

Comment:
There isn't really in this case since it's only for process and thread init, so I'll remove it for the sake of aesthetics. In general I think it would be a good practice to do a little loop unrolling where possible - e.g. in the memcpy stuff or anything performance-sensitive (which this is not of course).